### DISCUSSION TITLE
this.endComment is null

### DISCUSSION THREAD

**Discussion by TerrePorter** [livewire/livewire#6965] [7. Bugs (archived)]
### Livewire version

v3.0.3

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

I have trimmed down my page that I was having the error on to this very basic example. I created a test route and livewire component loading template.
```
Route::get('/testarray', function () {
    return view('testarray_index');
});
```

Testarray_index.blade.php
```
@extends('layouts.app')

@section('content')
    <div>
        @livewire('test-array')
    </div>
@endsection
```

That extends the layout which loads the livewire scripts. I am not using the autoload feature for livewire so I have the @livewireStyles and @livewireScripts tags in the layout file.

The test livewire component template.

```
<div>
    <div >
        <button wire:click="dotest">TEST</button>
    </div>
    <div class="p-4">
        @foreach($myRandomNamesData as $randomName)
           <p>
                @if ($randomName[1])
                    <a class="underline">
                        {{ $randomName[0] }}
                    </a>
                @else
                    <div>
                        {{ $randomName[0] }}
                    </div>
                @endif
            </p>
        @endforeach
    </p>
</div>
```

The test livewire component code.

```
<?php

namespace App\Livewire;

use Livewire\Component;

class TestArray extends Component
{
    public $myRandomNamesData;

    public function dotest() {
        $this->makeArray();
    }

    public function makeArray() {

        $myRandomNamesData = [];

        $myRandomNamesData[] = [
            'TEST ' . rand(1,100),
            rand(0,1)
        ];

        $myRandomNamesData[] = [
            'TEST ' . rand(1,100),
            rand(0,1)
        ];

        $myRandomNamesData[] = [
            'TEST ' . rand(1,100),
            rand(0,1)
        ];

        $myRandomNamesData[] = [
            'TEST ' . rand(1,100),
            rand(0,1)
        ];

        $this->myRandomNamesData = $myRandomNamesData;
    }

    public function render()
    {
        $this->makeArray();
        return view('livewire.test-array');
    }
}
```

The process is simply. Load the route and the data is loaded and displayed. Click the test button and the random data is reloaded and displayed.

The issue is when the page is reloaded after clicking the button an error message occurs.

```
Uncaught TypeError: this.endComment is null  livewire.js:6446:7
```

This only happens with the exact structure of template. If I change or remove the P tag the error does not happen. If I change or remove the A tag the error does not happen. Changes to the div tag in the foreach does not appear to make any difference. It only is the P tag and A tag that have an effect on the outcome.

Hope that helps reproduce the issue. 

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by MichaelPote** [comment]
I've also encountered this bug. For me I had the following:

```
		    @foreach($students as $student)
				@if (!empty($student->room_bed))
					    @continue
				@endif
						..... HTML .....
		    @endforeach
```

And when re-rendering the page after a change I would get `this.endComment is null` and Livewire would try to do masses of AJAX calls.
I fixed it by removing the `@continue` and inlining the test:

```
		    @foreach($students->filter(fn($student) => empty($student->room_bed)) as $student)
					.... HTML....
		    @endforeach
```

**Reply by DanielGSoftware** [reply]
This fixed it for us as well. Thanks!

**Comment by calebporzio** [comment]
Hmm. this probably has to do with Livewire's morph marker injection stuff.

Try disabling it in your livewire.php config: 

![CleanShot 2024-01-23 at 10 14 01@2x](https://github.com/livewire/livewire/assets/3670578/42852e11-ec41-4d67-9857-d3fcaf59c3ad)

If you set that to false, this might fix the problem.

Report back either way though

**Reply by DanielGSoftware** [reply]
Hi @calebporzio, not my thread, but the [solution](https://github.com/livewire/livewire/discussions/6965#discussioncomment-8219373) provided by @MichaelPote worked for us. When I disabled `inject_morph_markers`, it didn't fix it for us. However, in our specific case we had `@continue` in our code.

**Reply by A1exMiranda** [reply]
This actually fixed my error, thank you so much!

**Reply by joshhanley** [reply]
@DanielGSoftware ah interesting, looks like the morph markers should also be added for when `@continue` is used, as it's likely not allowing the morph markers to "finish" correctly as it would at the end of the loop. Any chance you could submit a failing test PR?

**Reply by montes** [reply]
Thanks! This also solved a similar problem for me (maybe you'll need a view:clear after updating the config)

**Reply by chris-arated** [reply]
This fixed it for me, only after view:clear (thanks @montes)

I'm still not sure why this started happening, I haven't changed a lot but the error was coming up when sorting my data table and rows were disappearing.

**Comment by Ahrengot** [comment]
I encountered the same issue when i was conditionally rendering a slot in a nested blade component. 

Turning off morph makers didn't fix it. 

What I had before:

```
<x-table>
    <x-slot:body>
        ...
    </x-slot:body>

    @if ($fuel->isNotEmpty())
        <x-slot:footer>
            <x-table.row class="border-t">
                ...
            </x-table.row>
        </x-slot:footer>
    @endif
</x-table>
```

The fix 
```
<x-table>
    <x-slot:body>
        ...
    </x-slot:body>

    <x-slot:footer>
        <x-table.row @class(['border-t', 'hidden' => $fuel->isEmpty()])>
            ...
        </x-table.row>
    </x-slot:footer>
</x-table>
```

**Reply by joshhanley** [reply]
@Ahrengot ah yeah the morph markers wouldn't know where to put the conditional markers as the slot spot in the blade component is different to where it's rendered in the parent. I would probably just put the conditional inside the slot.

```blade
<x-table>
    <x-slot:body>
        ...
    </x-slot:body>

    <x-slot:footer>
        @if ($fuel->isNotEmpty())
            <x-table.row class="border-t">
                ...
            </x-table.row>
         @endif
    </x-slot:footer>
</x-table>
```

**Comment by OmeBlues** [comment]
Also encountered this. I wanted to add a comma to list of (linked) authors but not on the last one. The comma was just added as is, encapsulating it inside a `<span>` did the trick for me.

**Comment by colbyalbo** [comment]
I also ran into this. I had 2 blade foreach loops in the same DIV element. 
Putting each  blade loop inside its own DIV remedied the error.

php 8.3
Laravel 11.7.0
Livewire 3.4.12

**Comment by morceaudebois** [comment]
I just updated my project from Livewire 2 to 3 and had a similar error: `Uncaught TypeError: this.endComment is undefined`

Turns out I had just forgotten to close a `<tr>` element:

```php
@foreach ($logs as $log)
    <tr>
         <td>
              @if (isset($log->user))
                   {{ $log->user->name }}
              @endif
         </td>
    </tr> // this one was missing 
@endforeach
```

**Reply by uandco** [reply]
+1, in my case a mismatched closing tag triggered it when the whole element was removed from the DOM because the condition was not met anymore. i.e.

```
@if ($something)
<custom-element>
</div>
@endif
```

As soon as `$something` was false, it would trigger the error, because I had forgotten to change the closing tag to `</custom-element>` when refactoring a bunch of markup.

**Reply by charliemerland** [reply]
Mismatched closing tag indeed, my editor mistakenly added a `</buttonclass>` tag that went unnoticed. Thank you both!

**Comment by jlopinto** [comment]
Be careful with nested forms! Browsers will transform:  

```html
<form id="form-1">
  <div>form 1</div>
  <form id="form-2"> 
    <div>form 2</div>
  </form>
</form>
```  

into  

```html
<form id="form-1">
  <div>form 1</div>
  <div>form 2</div>
</form>
```  

This may interfere with Livewire's DOM diffing.
